base/util.jl

```diff
@@ -313,6 +313,21 @@ end
 
 julia_exename() = ccall(:jl_is_debugbuild,Cint,())==0 ? "julia" : "julia-debug"
 
+"""
+    securezero!(o)
+
+`securezero!` fills the memory associated with an object `o` with zeros.
+Unlike `fill!(o,0)` and similar code, which might be optimized away by
+the compiler for objects about to be discarded, the `securezero!` function
+will always be called.
+"""
+function securezero! end
+@noinline securezero!{T<:Number}(a::AbstractArray{T}) = fill!(a, 0)
+securezero!(s::String) = securezero!(s.data)
+@noinline unsafe_securezero!{T}(p::Ptr{T}, len::Integer=1) =
+    ccall(:memset, Ptr{T}, (Ptr{T}, Cint, Csize_t), p, 0, len*sizeof(T))
```

Comment:
Also, by calling it `securezero!`, it gives an indication at the call site that something is being zeroed for security reasons.  This helps to prevent "human optimizers" from optimizing it out. 😉